daal4py/sklearn/neighbors/knn_classifier.py

```diff
@@ -0,0 +1,546 @@
+        if mode == 'connectivity':
+            A_ind = self.kneighbors(X, n_neighbors, return_distance=False)
+            n_queries = A_ind.shape[0]
+            A_data = np.ones(n_queries * n_neighbors)
+
+        elif mode == 'distance':
+            A_data, A_ind = self.kneighbors(
+                X, n_neighbors, return_distance=True)
+            A_data = np.ravel(A_data)
+
+        else:
+            raise ValueError(
+                'Unsupported mode, must be one of "connectivity" '
+                'or "distance" but got "%s" instead' % mode)
+
+        n_queries = A_ind.shape[0]
+        n_samples_fit = self.n_samples_fit_
+        n_nonzero = n_queries * n_neighbors
+        A_indptr = np.arange(0, n_nonzero + 1, n_neighbors)
+
+        kneighbors_graph = csr_matrix((A_data, A_ind.ravel(), A_indptr),
+                                      shape=(n_queries, n_samples_fit))
+
+        return kneighbors_graph
+
+
+class KNeighborsClassifier(NeighborsBase, KNeighborsMixin,
+                           SupervisedIntegerMixin, ClassifierMixin):
+
+    @_deprecate_positional_args
+    def __init__(self, n_neighbors=5, *,
+                 weights='uniform', algorithm='auto', leaf_size=30,
+                 p=2, metric='minkowski', metric_params=None, n_jobs=None,
+                 **kwargs):
+        super().__init__(
+            n_neighbors=n_neighbors,
+            algorithm=algorithm,
+            leaf_size=leaf_size, metric=metric, p=p,
+            metric_params=metric_params,
+            n_jobs=n_jobs, **kwargs)
+        self.weights = _check_weights(weights)
+
+    def predict(self, X):
+        """Predict the class labels for the provided data.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_queries, n_features), \
+                or (n_queries, n_indexed) if metric == 'precomputed'
+            Test samples.
+
+        Returns
+        -------
+        y : ndarray of shape (n_queries,) or (n_queries, n_outputs)
+            Class labels for each data sample.
+        """
+        X = check_array(X, accept_sparse='csr')
+
+        try:
+            fptype = getFPType(X)
+        except ValueError:
+            fptype = None
+
+        if self.weights in ['uniform', 'distance'] and self.algorithm in ['brute'] \
+        and (self.metric == 'minkowski' and self.p == 2 or self.metric == 'euclidean') \
+        and self._y.ndim == 1 and fptype is not None:
+
+            n_classes = len(self.classes_)
+
+            if self.algorithm == 'brute':
+                knn_classification_training = d4p.bf_knn_classification_training
+                knn_classification_prediction = d4p.bf_knn_classification_prediction
+            else:
+                knn_classification_training = d4p.kdtree_knn_classification_training
+                knn_classification_prediction = d4p.kdtree_knn_classification_prediction
+
+            training_alg = knn_classification_training(
```

Comment:
Are you sure that training for the KDTree should be done at the predict stage?

Reply:
It's a bit confusing but sklearn kNN training is performed on predict method.

Reply:
For KDTree?